src/openrct2/System.hpp

```diff
@@ -0,0 +1,60 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2021 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+namespace OpenRCT2
+{
+    struct IContext;
+
+    // Base class for systems that are hosted in the Context.
+    class System
+    {
+        IContext& _context;
+
+    public:
+        System(IContext& owner)
+            : _context(owner)
+        {
+        }
+
+        virtual ~System() = default;
+
+        // Called before a tick.
+        virtual void PreTick()
+        {
+        }
+
+        // Called every game tick, which by default is 40hz.
+        virtual void Tick()
+        {
+        }
+
+        // Called after a tick.
+        virtual void PostTick()
+        {
+        }
+
+        // Called every frame.
+        virtual void Update()
+        {
+        }
```

Comment:
(On `https://github.com/OpenRCT2/OpenRCT2/pull/15232/commits/1cfc933a59f37355ed8f5f614daf7e9429cb65b7`)

Does it make sense for those to have empty implementations or should they be abstract?

Reply:
Not all systems will need all of those events  thats why they are empty, this is future stuff.